python/requirements.in

```diff
@@ -0,0 +1,7 @@
+# This duplicates the requirements stated in setup.cfg because
```

Comment:
Can you put a similar comment in setup.cfg? Otherwise this is gonna get out of sync real fast. Also, we should be able to write a genrule to create this from setup.cfg; would that work?

Reply:
The genrule idea would have been a great way to solve this. Unfortunately, there is a blocker because `install_pip` is a `repository_rule` using `repository_context` and "remote repositories are executed during the analysis phase and thus cannot depends on a target result (the label should point to a non-generated file)" - see [docs](https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#path) and this [feature request](https://github.com/bazelbuild/bazel/issues/4069). I pointed to `#path` because that's where in the install it's breaking. 

Instead, I wrote in a pre-bazel dependency which you can see in the Makefile. I am definitely not in love with that, but to me it seemed slightly better than enforcing via comments. Let me know what you think.

Reply:
As someone who never uses the makefile, I'm also not in love with that.

Reply:
Other option: check in the `requirements.generated` file. Add a comment in both places reminding people to use the script to update the `requirements.generated` file when the `install_requires` changes. 
Only problem: the end library consumers would also see that comment and be confused. But likely worthwhile price to pay.

Reply:
to summarize:
- setup.cfg needs to be source of truth to work like a normal python package
- setup.cfg can't be consumed by bazel directly. And also bazel can't consume generated targets in its place (at least not with @rules_python, and migrating to something less official isn't worth it)
- so we either keep the file it can consume in sync with setup.cfg manually or through a pre-bazel make rule. 

Defer to you here.

Reply:
Thanks, that succinctly lays out the constraints. Here's what I'm thinking:
- having more than one source of truth isn't ideal, but is manageable if they're kept in sync
- keeping files in sync manually is difficult since it requires everyone making and reviewing changes to know the convention
- (as you suggest) we can make this less difficult and more automated with tooling

Bazel is more familiar to me and is well-integrated with CI; since `make` invokes Bazel and not the other way around, I'd rather make sure things stay in sync there. Since we can't generate the `requirements.txt` file within Bazel, the next best thing seems to be a test that confirms the contents are synchronized. That, at least, lets us make sure at CI time that things aren't broken. Also comments in both files pointing at the other to help out human editors.

Reply:
Done. There is now simply a bazel py_test to ensure the two files stay in sync. Agree that this is much better.